cunumeric/array.py

```diff
@@ -3113,32 +3122,46 @@ def mean(
                 dtype = np.dtype(np.float64)
             else:
                 dtype = self.dtype
+
+        where_array = broadcast_where(where, self.shape)
         # Do the sum
-        if out is not None and out.dtype == dtype:
-            sum_array = self.sum(
-                axis=axis,
-                dtype=dtype,
-                out=out,
-                keepdims=keepdims,
+        sum_array = (
+            self._nansum(
+                axis=axis, out=out, keepdims=keepdims, where=where_array
             )
-        else:
-            sum_array = self.sum(
-                axis=axis,
-                dtype=dtype,
-                keepdims=keepdims,
+            if out is not None and out.dtype == dtype and ignore_nan
+            else self.sum(
+                axis=axis, out=out, keepdims=keepdims, where=where_array
             )
+            if out is not None and out.dtype == dtype
+            else self._nansum(axis=axis, keepdims=keepdims, where=where_array)
+            if ignore_nan
+            else self.sum(axis=axis, keepdims=keepdims, where=where_array)
+        )
+
         if axis is None:
-            divisor = reduce(lambda x, y: x * y, self.shape, 1)
+            if where_array is not None:
+                divisor = np.array(where_array._count_nonzero())
+            else:
+                divisor = np.array(reduce(lambda x, y: x * y, self.shape, 1))
+
         else:
-            divisor = self.shape[axis]
+            if where_array is not None:
+                divisor = np.array(
+                    where_array.sum(axis=axis, keepdims=keepdims)
+                )
```

Comment:
`where_array.sum` will sometimes return an integer, so we souldn't be able to run `astype` in the `__itruediv__` 
Bit I think I will just add `dtype` to the sum instead of calling `astype` later.